proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,383 @@
+//! State machine to manage the current mode of operation during peer lifecycle.
+
+use std::{
+    collections::HashSet,
+    time::{Duration, Instant},
+};
+
+use librad::{
+    net::{peer::Gossip, protocol::ProtocolEvent},
+    peer::PeerId,
+};
+
+use crate::peer::announcement;
+
+/// Default number of peers a full sync is attempting with up on startup.
+/// TODO(xla): Revise number.
+const DEFAULT_SYNC_MAX_PEERS: usize = 5;
+
+/// Default Duration until the local peer goes online regardless if and how many syncs have
+/// succeeded.
+// TODO(xla): Review duration.
+const DEFAULT_SYNC_PERIOD: Duration = Duration::from_secs(5);
+
+/// Instructions to issue side-effectful operations which are the results from state transitions.
+#[derive(Debug, PartialEq)]
+pub enum Command {
+    /// Start the announcement subroutine.
+    Announce,
+    /// Initiate a full sync with `PeerId`.
+    SyncPeer(PeerId),
+    /// Start sync timeout.
+    StartSyncTimeout(Duration),
+}
+
+/// Significant events that occur during [`Peer`] lifetime.
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Debug)]
+pub enum Event {
+    /// Announcement subroutine lifecycle events.
+    Announce(AnnounceEvent),
+    /// Events from the underlying coco protocol.
+    Protocol(ProtocolEvent<Gossip>),
+    /// Lifecycle events during peer sync operations.
+    PeerSync(SyncEvent),
+    /// Scheduled timeouts which can occur.
+    Timeout(TimeoutEvent),
+}
+
+/// Announcement subroutine lifecycle events.
+#[derive(Clone, Debug)]
+pub enum AnnounceEvent {
+    /// Operation failed.
+    Failed,
+    /// Operation succeeded and emitted the enclosed list of updates.
+    Succeeded(announcement::Updates),
+    /// The ticker duration has elapsed.
+    Tick,
+}
+
+/// Lifecycle events during peer sync operations.
+#[derive(Clone, Debug)]
+pub enum SyncEvent {
+    /// A sync has been initiated for `PeerId`.
+    Started(PeerId),
+    /// A sync has failed for `PeerId`.
+    Failed(PeerId),
+    /// A sync has succeeded for `PeerId`.
+    Succeeded(PeerId),
+}
+
+/// Scheduled timeouts which can occur.
+#[derive(Clone, Debug)]
+pub enum TimeoutEvent {
+    /// Grace period is over signaling that we should go offline, no matter how many syncs have
+    /// succeeded.
+    SyncPeriod,
+}
+
+/// The current status of the local peer and its relation to the network.
+#[derive(Debug)]
+enum Status {
+    /// Nothing is setup, not even a socket to listen on.
+    Stopped(Instant),
+    /// Local peer is listening on a socket but has not connected to any peers yet.
+    Started(Instant),
+    /// The local peer lost its connections to all its peers.
+    Offline(Instant),
+    /// Phase where the local peer tries get up-to-date.
+    Syncing(Instant, usize),
+    /// The local peer is operational and is able to interact with the peers it has connected to.
+    Online(Instant),
+}
+
+/// Set of knobs to change the behaviour of the [`RunState`].
+pub struct Config {
```

Comment:
`nit` As far as I see, the only use case for `Config` is to support the syncing stage, thus I'd suggest:

```suggestion
pub struct SyncConfig {
```

+ and drop "sync_" as prefix from all fields. leave it up to you :ping_pong: 
+ move it to `sync.rs`

Up to you :raised_hands:

Reply:
Yeah not a bad suggestion, just that this struct will be used to house all run loop configuration. Good candidate is the announcement tick duration and other knobs in the future. What could be done is to have separate struct bundling all sync concerns that is a field of `Config`.